torch/lib/THCUNN/LookupTable.cu

```diff
@@ -162,36 +162,76 @@ __global__ void cunn_LookupTable_accGradParametersKernel(
   }
 }
 
-/*
- * Keep the norm of weight smaller than maxNorm
- */
 template <typename Dtype, typename Acctype>
-struct pow_v
-{
-  Acctype normType;
-  pow_v(Dtype v) : normType(ScalarConvert<Dtype, Acctype>::to(v)) {}
-  __host__ __device__
-  Acctype operator()(const Dtype& x) const {
+__device__
+Acctype fast_pow(Dtype x, Acctype normType) {
     Acctype xA = ScalarConvert<Dtype, Acctype>::to(x);
     if (normType == 1)
       return std::abs(xA);
     else if (normType == 2)
       return xA * xA;
     else
       return std::pow(std::abs(xA), normType);
-  }
-};
+}
+
+template <typename Dtype>
+__device__
+Dtype get_padded_value(Dtype *weight_ptr, THCIndex_t idx, int64_t dim) 
+{
+  if (idx < dim)
+    return weight_ptr[idx];
+  else
+    return 0;
+}
 
-template <typename T>
-struct multiply_s
+/* Calculate norms of the rows of weight_ptr given by idx_ptr and capture them in norms */
+template <typename Dtype, typename Acctype>
+__global__
+void calculate_norms_and_renorm(Dtype *weight_ptr, THCIndex_t *idx_ptr, Dtype normType, 
+    Dtype maxNorm, Acctype normType_, int64_t dim)
 {
-  T scale;
-  multiply_s(T s) : scale(s) {}
-  __host__ __device__
-  T operator()(const T& x) const {
-    return x * scale;
+  // Some casting hacks since dynamic shared memory and templates don't work together:
+  extern __shared__ __align__(sizeof(Acctype)) unsigned char smem[];
+  Acctype *sdata = reinterpret_cast<Acctype *>(smem);
+
+  int64_t tid = threadIdx.x;
+  int64_t idx = (idx_ptr[blockIdx.x] - TH_INDEX_BASE) * dim + tid;
+
+  // Do the initial exponents in the load
+  // Shenanigans here since we need to work with a power-of-2 block size
+  // effectively pad the matrix with zeroes if needed
+  Acctype second;
+  if (tid + blockDim.x > dim)
```

Comment:
put brackets for all if/for/while/etc. statements

Reply:
also, for loading with a fixed block size and with the block reduction function below, just use a block-stride loop to sum locally and then reduce.

Also, don't use the constants "0" and "1" etc., which won't work if this code is made to work on half or other types. Use THCNumerics, ScalarConvert etc. from cutorch instead. pow() should use that too:

```
AccType v = ScalarConvert<int, AccType>::to(0);
for (IndexType i = threadIdx.x; i < dim; i += blockDim.x) {
  v += FastPow<DType, AccType, Norm>::pow(weight[i]);
}

v = reduceBlock<AccType, WhateverTheSumOpIs>(smem, ...);

if (threadIdx.x == 0) {
  sdata[0] = std::pow(v, THCNumerics<AccType>::div(ScalarConvert<int, Acctype>::to(1),
             ScalarConvert<int, Acctype>::to(Norm));
}
```